pkg/agent/stats/collector.go

```diff
@@ -163,12 +177,80 @@ func addUp(stats *statsv1alpha1.TrafficStats, inc *agenttypes.RuleMetric) {
 	stats.Bytes += int64(inc.Bytes)
 }
 
+func isIdenticalMulticastGroupMap(a, b map[string][]cpv1beta.PodReference) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for aKey, aValue := range a {
+		bValue, exist := b[aKey]
+		if !exist {
+			return false
+		}
+		if len(aValue) != len(bValue) {
+			return false
+		}
+		aValueSet := sets.NewString()
+		for _, av := range aValue {
+			aValueSet.Insert(k8s.NamespacedName(av.Name, av.Namespace))
+		}
+		bValueSet := sets.NewString()
+		for _, bv := range bValue {
+			bValueSet.Insert(k8s.NamespacedName(bv.Name, bv.Namespace))
+		}
+		if !aValueSet.Equal(bValueSet) {
+			return false
+		}
+	}
+	return true
+}
+
 // report calculates the delta of the stats and pushes it to the antrea-controller summary API.
+// If multicast feature gate is enabled, it also send the full multicast group and IGMP report stats to the antrea-controller.
 func (m *Collector) report(curStatsCollection *statsCollection) error {
 	npStats := calculateDiff(curStatsCollection.networkPolicyStats, m.lastStatsCollection.networkPolicyStats)
 	acnpStats := calculateRuleDiff(curStatsCollection.antreaClusterNetworkPolicyStats, m.lastStatsCollection.antreaClusterNetworkPolicyStats)
 	anpStats := calculateRuleDiff(curStatsCollection.antreaNetworkPolicyStats, m.lastStatsCollection.antreaNetworkPolicyStats)
-	if len(npStats) == 0 && len(acnpStats) == 0 && len(anpStats) == 0 {
+
+	multicastGroups := []cpv1beta.MulticastGroupInfo{}
+	multicastGroupsUpdated := false
+	if m.multicastEnabled {
+		// multicastGroups should be reported if the multicast group Pod membership has changed since the last collect.
+		if !isIdenticalMulticastGroupMap(m.lastStatsCollection.multicastGroups, curStatsCollection.multicastGroups) {
+			multicastGroupsUpdated = true
+			multicastGroups = make([]cpv1beta.MulticastGroupInfo, 0, len(curStatsCollection.multicastGroups))
+			for group, pods := range curStatsCollection.multicastGroups {
+				multicastGroups = append(multicastGroups, cpv1beta.MulticastGroupInfo{Group: group, Pods: pods})
+			}
+		}
+
+		// Collect statistics of IGMP report messages hit by ANP or ACNP, and merge them to anpStats and acnpStats.
```

Comment:
Note that if a report failed, other metrics can still reported to antrea-controller later as it doesn't update `m.lastStatsCollection` on failure, but IGMP metrics would be lost. Even there is no good solution for now, we should note down the defect.

Reply:
Comment updated. Thanks for mentioning that point.

Reply:
Comment updated. Thanks for mentioning that point.